src/instructlab/cli/model/upload.py

```diff
@@ -0,0 +1,245 @@
+# SPDX-License-Identifier: Apache-2.0
+
+# Standard
+from pathlib import Path
+import abc
+import logging
+import os
+
+# Third Party
+from huggingface_hub import HfApi
+from huggingface_hub import logging as hf_logging
+import click
+
+# First Party
+from instructlab import clickext
+from instructlab.configuration import DEFAULTS
+from instructlab.defaults import DEFAULT_INDENT
+from instructlab.utils import is_model_gguf, is_model_safetensors
+
+logger = logging.getLogger(__name__)
+
+
+class ModelUploader(abc.ABC):
+    """Base class for a model uploading backend"""
+
+    def __init__(
+        self,
+        ctx,
+        model: str,
+        model_dir: str,
+        upload_dest: str,
+        release: str,
+    ) -> None:
+        self.ctx = ctx
+        self.model = model
+        self.model_dir = model_dir
+        self.upload_dest = upload_dest
+        self.release = release
+
+    @abc.abstractmethod
+    def upload(self) -> None:
+        """Uploads specified local model and stores it into upload_dest@release"""
+
+
+class HFModelUploader(ModelUploader):
+    """Class to handle uploading safetensors and GGUF models to Hugging Face"""
+
+    def __init__(
+        self,
+        ctx,
+        model: str,
+        model_dir: str,
+        upload_dest: str,
+        release: str,
+        hf_token: str,
+    ) -> None:
+        super().__init__(
+            ctx=ctx,
+            model=model,
+            model_dir=model_dir,
+            upload_dest=upload_dest,
+            release=release,
+        )
+        self.hf_token = hf_token
+        self.local_model_path = ""
+
+    def upload(self):
+        """
+        Upload specified model to Hugging Face
+        """
+
+        # HF token check
+        if self.hf_token == "":
+            raise ValueError(
+                """HF_TOKEN var needs to be set in your environment to upload HF Model.
+                Alternatively, the token can be passed with --hf-token flag.
+                The HF Token is used to authenticate your identity to the Hugging Face Hub."""
+            )
```

Comment:
Is it possible someone could pass in a whitespace token like `"   "`? I suggest we use `.strip()` to strip all the whitespaces

We also might want to check for the `None` case

```suggestion
        # HF token check
        if not isinstance(self.hf_token, str):
            raise ValueError(
                """The HF Token is either unset or not a valid string.
                Please ensure that either the HF_TOKEN var is set to a valid string in your environment
                or pass a valid HF token string using the --hf-token flag.
                The HF Token is used to authenticate your identity to the Hugging Face Hub."""
            )
        if len(self.hf_token.strip()) == 0:
            raise ValueError(
                """HF_TOKEN var needs to be set in your environment to upload HF Model.
                Alternatively, the token can be passed with --hf-token flag.
                The HF Token is used to authenticate your identity to the Hugging Face Hub."""
            )
```

Reply:
I believe (@cdoern can confirm) the string check is already done by Click 

However, I will go ahead and do the strip/len check for the case you mentioned.

We may want to upload the `ilab model download` check in a separate PR as well?

Reply:
I think you're right about the string check. The `click` docs say:

> If no type is provided, the type of the default value is used. If no default value is provided, the type is assumed to be [STRING](https://click.palletsprojects.com/en/stable/api/#click.STRING).

Source: https://click.palletsprojects.com/en/stable/options/#basic-value-options

Since `default=""`, which is obviously a `str`, it seems `click` will assume the variable's type is `str`.

```suggestion
@click.option(
    "--hf-token",
    default="",
    envvar="HF_TOKEN",
    type=str,
    help="User access token for connecting to the Hugging Face Hub.",
)
```